### PR TITLE
WMT PyTorch Workload

### PR DESCRIPTION
**Pull request by runame**
This PR adds
- WMT PyTorch workload #29 (WIP). The convergence of the loss/accuracy matches the Jax implementation in terms of iterations.
- Implements the `param_shapes` property for the PyTorch MNIST, ImageNet, and WMT workload. As @fsschneider has suggested before, I think properties like this should be placed in a general parent class, something like `PytorchWorkload`.
- `test_num_params`, which checks if the number of model parameters of the Jax and PyTorch versions of the models from the MNIST, ImageNet, and WMT workloads match.
- `test_param_shapes`, which uses the `param_shapes` property of the MNIST, ImageNet, and WMT workload to check if the number of parameter tensors match. Not ideal yet, see my [comment](https://github.com/runame/algorithmic-efficiency/blob/wmt-pytorch/tests/test_param_shapes.py#L21-L24) in the code. Addresses #53.
- PyTorch CPU install option (mostly for `pytest`) and `tensorflow-text` requirement for WMT workload.

Most important TODOs to finish the workload:
- [x] Support multiple GPUs for decoding.
- [ ] Test the implementation of the cache for autoregressive decoding (verify that bleu score matches Jax implementation).
- [ ] Speed up training (less time per iteration), e.g. by using `DistributedDataParallel` instead of `DataParallel`.
- [x] Add metrics on subset of training data (also not implemented for the Jax version of the workload yet).

### CODE REVIEW

**Review by znado — APPROVED** [review]
overall looks awesome, had a few minor comments. feel free to merge whenever you're ready, I think we can solve the remaining TODOs in smaller follow-up PRs

File: algorithmic_efficiency/workloads/wmt/wmt_pytorch/models.py [hunk 1]
```diff
@@ -0,0 +1,1208 @@
...
+  src_mask = make_attention_mask(src > 0, src > 0)
+  # Add segmentation block-diagonal attention mask if using segmented data.
+  if inputs_segmentation is not None:
+    src_mask = torch.logical_and(
+        src_mask,
+        make_attention_mask(inputs_segmentation, inputs_segmentation, torch.eq))
+  # Flip values and ensure numerical stability.
+  src_mask = torch.repeat_interleave(
+      torch.logical_not(src_mask), repeats=nhead, dim=0)
+  new_src_mask = torch.zeros_like(src_mask, dtype=torch.float32)
```
[the remark this answers is not in the document]
Comment (runame) [reply algorithmic_efficiency/workloads/wmt/wmt_pytorch/models.py:69]: "For a float mask, the mask values will be added to the attention weight." ([source](https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html#torch.nn.MultiheadAttention)) So `zeros_like` is correct here.

File: algorithmic_efficiency/workloads/wmt/wmt_pytorch/workload.py [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+
+
+def delete_decoding_cache(model):
+  for module in model.modules():
+    if len(list(module.buffers(recurse=False))) > 0:
+      names = [name for name, _ in module.named_buffers() if name != 'pe']
+      # Deletion has to be done in seperate for-loop to avoid
+      # mutating the OrderedDict during iteration.
+      for name in names:
+        del module._buffers[name]
```
[the remark this answers is not in the document]
Comment (runame) [reply algorithmic_efficiency/workloads/wmt/wmt_pytorch/workload.py:38]: `jax_to_pytorch` and `pytorch_to_jax` might be useful for other workloads as well, but when that is the case we can still put them in a separate utils file. So for now I'll just follow your suggestion.

File: algorithmic_efficiency/workloads/wmt/wmt_pytorch/workload.py [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+    vocab_size = logits.shape[-1]
+    confidence = 1.0 - label_smoothing
+    low_confidence = (1.0 - confidence) / (vocab_size - 1)
+    normalizing_constant = -(
+        confidence * np.log(confidence) +
+        (vocab_size - 1) * low_confidence * np.log(low_confidence + 1e-20))
+    one_hot_targets = F.one_hot(targets, num_classes=vocab_size)
+    soft_targets = torch.where(one_hot_targets == 1, confidence, low_confidence)
+    loss = super().forward(
+        input=logits.transpose(-2, -1), target=soft_targets.transpose(-2, -1))
```
[the remark this answers is not in the document]
Comment (runame) [reply algorithmic_efficiency/workloads/wmt/wmt_pytorch/workload.py:53]: I have only created a separate class to inherit from `nn.CrossEntropyLoss` because `DataParallel` is typically applied to modules, although I have just noticed that there is also a [function](https://pytorch.org/docs/stable/generated/torch.nn.functional.torch.nn.parallel.data_parallel.html) for that. It doesn't work with `DistributedDataParallel` though so I will leave this as a separate module for now and revisit it when I move away from `DataParallel`.

File: algorithmic_efficiency/workloads/wmt/input_pipeline.py [hunk 4]
```diff
@@ -292,16 +292,21 @@ def filter_fn(x):
...
+                     vocab_path: Optional[str] = None,
+                     pack_examples: bool = True):
   """Load and return dataset of batched examples for use during training."""
   if vocab_path is None:
     vocab_path = os.path.expanduser('~/wmt_sentencepiece_model')
 
-  train_ds_builder = tfds.builder('wmt17_translate/de-en')
+  train_ds_builder = tfds.builder(
+      'wmt17_translate/de-en',
+      data_dir='/mnt/qb/work/hennig/reschenhagen26/tensorflow_datasets/')
```
[the remark this answers is not in the document]
Comment (runame) [reply algorithmic_efficiency/workloads/wmt/input_pipeline.py:303]: Ah I just forgot about that.

File: algorithmic_efficiency/workloads/imagenet/imagenet_pytorch/workload.py [hunk 5]
```diff
@@ -34,10 +34,15 @@ class ImagenetWorkload(BaseImagenetWorkload):
...
-    raise NotImplementedError
+    """Return shape tuples from model as a tree."""
+    model, _ = self.init_model_fn([0])
+    param_shapes = {}
+    for name, module in model.named_modules():
+      if len(list(module.parameters(recurse=False))) > 0:
+        param_shapes[name] = {}
+        for param_name, param in module.named_parameters(recurse=False):
+          param_shapes[name][param_name] = spec.ShapeTuple(param.shape)
+    return param_shapes
```
[the remark this answers is not in the document]
Comment (runame) [reply algorithmic_efficiency/workloads/imagenet/imagenet_pytorch/workload.py:45]: I have created an issue (#74) for that and will do it in a separate PR, to check for functions in all workloads which should be put in a shared place.

File: algorithmic_efficiency/workloads/wmt/wmt_pytorch/workload.py [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+      predicted = self.predict_step(inputs,
+                                    params,
+                                    decode.EOS_ID,
+                                    max_predict_length)
+
+      # Iterate through non-padding examples of batch.
+      for i, s in enumerate(predicted[:cur_pred_batch_size]):
+        sources.append(decode_tokens(inputs[i]))
+        references.append(decode_tokens(targets[i]))
+        predictions.append(decode_tokens(s))
```
[the remark this answers is not in the document]
Comment (runame) [reply algorithmic_efficiency/workloads/wmt/wmt_pytorch/workload.py:166]: Maybe, have to look at the Jax code for that, but I don't think it matters much, so I will check this later, after the important leftover TODOs are done.

File: algorithmic_efficiency/workloads/wmt/wmt_pytorch/workload.py [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+
+    _, bleu_score = self.translate_and_calculate_bleu(
+        params=params,
+        predict_ds=self._predict_ds,
+        decode_tokens=self._decode_tokens,
+        max_predict_length=256)
+
+    eval_results['bleu'] = bleu_score
+
+    return eval_results
```
[the remark this answers is not in the document]
Comment (runame) [reply algorithmic_efficiency/workloads/wmt/wmt_pytorch/workload.py:328]: Makes sense to me, added it to the TODOs.

File: tests/test_num_params.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
+import jax
+import jax.numpy as jnp
+import jax.random as jax_rng
+import pytest
```
[the remark this answers is not in the document]
Comment (runame) [reply tests/test_num_params.py:4]: I can do that. I'm curious why you prefer `absl.testing`, given that `pytest` is already a requirement anyway?

File: algorithmic_efficiency/workloads/wmt/wmt_pytorch/models.py [hunk 9]
```diff
@@ -0,0 +1,1208 @@
...
+          self.register_buffer(name, c)
+
+    if self.batch_first and is_batched:
+      return attn_output.transpose(1, 0), attn_output_weights
+    else:
+      return attn_output, attn_output_weights
+
+
+# Modified to create cache for autoregressive decoding.
+def multi_head_attention_forward(
```
[the remark this answers is not in the document]
Comment (runame) [reply algorithmic_efficiency/workloads/wmt/wmt_pytorch/models.py:830]: Agreed, just pruned some of the code.

File: .github/workflows/CI.yml [hunk 10]
```diff
@@ -16,7 +16,11 @@ jobs:
...
         pip install pytest
         pip install .[jax_cpu]
+        pip install .[pytorch_cpu]
+        pip install .[wmt]
     - name: Run pytest
       run: |
         pytest -vx tests/version_test.py
         pytest -vx tests/workloads/imagenet/imagenet_jax/workload_test.py
+        pytest -vx tests/test_num_params.py
+        pytest -vx tests/test_param_shapes.py
```
Comment (fsschneider) [inline .github/workflows/CI.yml:26]: Can we not just call something like `pytest -vx tests/`?
If we want to exclude certain tests it might be best to mark them as slow (e.g. https://docs.pytest.org/en/6.2.x/mark.html)

**Review by fsschneider — APPROVED** [review]
Looks good!
I just gave it a brief review, since @znado already reviewed it carefully.

I think this is ready to merge. What is still missing is to match the `iterations per time` to the JAX version but I think @runame can tackle this in a separate PR.

File: .github/workflows/CI.yml (hunk 10 above)
Comment (runame) [reply .github/workflows/CI.yml:26]: Cool, I didn't know about this functionality. I think Zack wanted to exclude the slow tests from being run by GitHub workflows for now.
